[c10d] dump from one and only one thread (PG0's monitor thread)

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120893

Summary:
When there are multiple PGs in a process and a hardware failure happens,
we found that multiple PGs/ threads in the same
process are competing to dump the same records at the same time. The
affects the reliability of dumps.

In this PR, we will try to make the change such that only one thread/PG
could dump: PG0's monitor thread. We use a static variable to indicate
that something (e.g., collective timeout) has triggered the dump
locally.

monitor thread would dump debug info under any one of the 3 conditions:
1: this static variable is set to true by the watchdog thread when it detects
a timeout or pipe dump signal
2: timeout signal is received from other ranks through tcpstore
3: no heartbeat of watchdog
Test Plan:
python test/distributed/test_c10d_nccl.py -k
test_timeout_dumps_on_stuck_ranks

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

shuqiangzhang committed:
[c10d] dump from one and only one thread (PG0's monitor thread)
Summary:
When there are multiple PGs in a process and a hardware failure happens,
we found that mulltiple PGs/ threads in the same
process are competing to dump the same records at the same time. The
affects the reliability of dumps.

In this PR, we will try to make the change such that only one thread/PG
could dump: PG0's monitor thread. We use a static variable to indicate
that something (e.g., collective timeout) has triggered the dump
locally.

monitor thread would dump debug info under any one of the 3 conditions:
1: this static variable is set to true by the watchdog thread when it detects
a timeout of pipe dump signal
2: timeout signal is received from other ranks through tcpstore
3: no heartbeat of watchdog
Test Plan:
python test/distributed/test_c10d_nccl.py -k
test_timeout_dumps_on_stuck_ranks

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] dump from one and only one thread (PG0's monitor thread)"


Summary:
When there are multiple PGs in a process and a hardware failure happens,
we found that mulltiple PGs/ threads in the same
process are competing to dump the same records at the same time. The
affects the reliability of dumps.

In this PR, we will try to make the change such that only one thread/PG
could dump: PG0's monitor thread. We use a static variable to indicate
that something (e.g., collective timeout) has triggered the dump
locally.

monitor thread would dump debug info under any one of the 3 conditions:
1: this static variable is set to true by the watchdog thread when it detects
a timeout of pipe dump signal
2: timeout signal is received from other ranks through tcpstore
3: no heartbeat of watchdog
Test Plan:
python test/distributed/test_c10d_nccl.py -k
test_timeout_dumps_on_stuck_ranks

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1553,21 +1580,11 @@ void ProcessGroupNCCL::watchdogHandler() {
+            if (dumpOnTimeout_) {
+              // signal the monitor thread to start dumping
+              shouldDump_.store(true);
+              // This sleep is used to give time for dumping before throwing
```
i think we could also consider never throwing exceptions or commAborting from watchdog thread, and, using a similar means (perhaps storing a static exception ptr if we think of a better way) letting the monitor thread assume sole responsibility for 'exiting the process' if it's signaled to do so

could be a separate PR. wdyt?

shuqiangzhang replied:
Yes, probably a single place to exit as well, will try it in another PR

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1197,6 +1200,32 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+            "this can be caused by GIL deadlock or other reasons such as network errors or ",
+            "bugs in the communications library (e.g. NCCL), etc. We tried our best to ",
+            "dump the debug info into the storage to help you debug the issue.");
+        break;
```
does this break mean 'exit the monitor loop and kill the job?'

bc that isn't safe.  When the 'dumpPipe' is invoked, we expect the dump signal to propagate here and dump, but the training should continue.  Only if the dump is caused by timeout we want to begin the teardown.

shuqiangzhang replied:
Oh, didn't know dumpPipe is for dump and continue

wconstab replied:
yea. its probably worth adding a test that after dumpPipe is written the process doesn't die. i thought we had that actually.

shuqiangzhang replied:
We have the test, but I misunderstood the dump logic, so to disable the Kill part through: os.environ['TORCH_NCCL_ENABLE_MONITORING'] = '0', in the next version, I will remove this env change

shuqiangzhang committed:
Update on "[c10d] dump from one and only one thread (PG0's monitor thread)"


Summary:
When there are multiple PGs in a process and a hardware failure happens,
we found that multiple PGs/ threads in the same
process are competing to dump the same records at the same time. The
affects the reliability of dumps.

In this PR, we will try to make the change such that only one thread/PG
could dump: PG0's monitor thread. We use a static variable to indicate
that something (e.g., collective timeout) has triggered the dump
locally.

monitor thread would dump debug info under any one of the 3 conditions:
1: this static variable is set to true by the watchdog thread when it detects
a timeout or pipe dump signal
2: timeout signal is received from other ranks through tcpstore
3: no heartbeat of watchdog
Test Plan:
python test/distributed/test_c10d_nccl.py -k
test_timeout_dumps_on_stuck_ranks

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1406,59 +1443,6 @@ std::string ProcessGroupNCCL::getNCCLWatchdogDebugInfo() {
 }
 
-#if defined(__linux__)
-struct DumpPipe {
```
Move the declaration to the header file so that heartbeatMonitor() function can access it, defining in cpp file will make the access depending on the order of function implementation

wconstab reviewed (approved):
lgtm. thanks!

shuqiangzhang commented:
@pytorchbot merge
